### PR TITLE
feat: add check dependency function to dataset service

### PR DESCRIPTION
**Pull request by sevu1**
Issue #, if available:

Description of changes: add optional check dependency function to dataset service

Checklist:

* [X] Have you successfully deployed to an AWS account with your changes?
* [X] Have you written new tests for your core changes, as applicable?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: workbench-core/datasets/src/dataSetService.ts [hunk 1]
```diff
@@ -123,9 +123,17 @@ export class DataSetService {
   /**
    * Removes a DataSet from the solution however it does not delete the storage.
    * @param dataSetId - the ID of the DataSet to remove.
+   * @param checkDependency - function to validate whether all required prerequisites are met before removing dataset.
```
Comment (bhudnell) [inline workbench-core/datasets/src/dataSetService.ts:126]: Mention that the function needs to throw an error if a prerequisite isnt met

File: workbench-core/datasets/src/dataSetService.test.ts [hunk 2]
```diff
@@ -297,11 +297,15 @@ describe('DataSetService', () => {
...
+      metaPlugin.removeDataSet = jest.fn();
       service = new DataSetService(audit, log, metaPlugin);
     });
 
     it('returns nothing when the dataset is removed', async () => {
-      await expect(service.removeDataSet(mockDataSetId)).resolves.not.toThrow();
+      await service.removeDataSet(mockDataSetId);
+
+      expect(metaPlugin.removeDataSet).toBeCalledTimes(1);
+      expect(metaPlugin.removeDataSet).toBeCalledWith(mockDataSetId);
```
Comment (bhudnell) [inline workbench-core/datasets/src/dataSetService.test.ts:308]: We should be testing that were getting the expected outputs with specific inputs, not implementation details. If we change how the function works at a later date, these tests will likely fail even if the expected behavior is the same

File: workbench-core/datasets/src/dataSetService.test.ts [hunk 3]
```diff
@@ -310,6 +314,18 @@ describe('DataSetService', () => {
           'External endpoints found on Dataset must be removed before DataSet can be removed.'
         )
       );
+
+      expect(metaPlugin.removeDataSet).toBeCalledTimes(0);
```
Comment (bhudnell) [inline workbench-core/datasets/src/dataSetService.test.ts:318]: same comment as above

File: workbench-core/datasets/src/dataSetService.test.ts [hunk 4]
```diff
@@ -310,6 +314,18 @@ describe('DataSetService', () => {
...
+    });
+
+    it('throws when preconditions are not met', async () => {
+      await expect(
+        service.removeDataSet(mockDataSetId, async () => {
+          await Promise.reject(new Error('Preconditions are not met'));
+        })
+      ).rejects.toThrow('Preconditions are not met');
+
+      expect(metaPlugin.removeDataSet).toBeCalledTimes(0);
```
Comment (bhudnell) [inline workbench-core/datasets/src/dataSetService.test.ts:328]: same comment as above

File: workbench-core/datasets/src/dataSetService.test.ts (hunk 2 above)
Comment (sevu1) [reply workbench-core/datasets/src/dataSetService.test.ts:308]: It's debatable - [white box testing](https://en.wikipedia.org/wiki/White-box_testing) where implementation details are known vs [black box testing](https://en.wikipedia.org/wiki/Black-box_testing) where we only know input and output.

**Review by nguyen102 — APPROVED** [review]

**Review by bhudnell — APPROVED** [review]

**Pull request merged**